### ISSUE TITLE
Disable B018 rule for Marimo notebooks?

### ISSUE THREAD

**Issue by wlach** [astral-sh/ruff#21301]
### Summary

Would it be possible to have the same behaviour for [B018](https://docs.astral.sh/ruff/rules/useless-expression/) as we do with ipynb cells for marimo notebooks? Adding a variable to display is pretty common at the end of a cell. As we've been using marimo more and more, this increasingly has become annoying. Obviously you can ignore it on a per-file/directory basis, but that's a tedious process that has to be done over and over as new directories or projects are added to our monorepo.

Marimo notebooks have a fairly deterministic structure that is relatively easy to detect:

```py
import marimo

__generated_with = "0.17.6"
app = marimo.App(width="medium")


@app.cell
def _():
    # This cell just for imports
    import pandas as pd
    return pd,


@app.cell
def _(pd):
    # 1. Create a silly pandas DataFrame
    data = {"col1": [1, 2], "col2": [3, 4]}
    df = pd.DataFrame(data)

    # 2. Display the DataFrame
    # This is the standard notebook pattern for cell output.
    # A linter will incorrectly flag this line as a "useless statement" (B018).
    df
    return


if __name__ == "__main__":
    app.run()
```

**Comment by dylwil3** [comment]
Interesting! Can you say a bit more about the precise situation we would ignore here?

Is it something like, all of the following hold for a given expression statement X?

1. X is the last top-level statement in the body of a function `f` excluding `return` statements
2. X is a "useless" expression (as in the usual rule logic)
3. The function `f` is decorated with `foo.cell`
4. `foo` is an instance of the class with qualified name `marimo.App`

In particular for (1), does the `return` statement have to be empty or should the lint rule fire if you're returning something and the preceding statement is a "useless expression"?

(I'm not immediately sold on implementing this, but I want to make sure I understand what it would look like)

**Comment by ntBre** [comment]
I think to some extent this is a request for Marimo notebook support. I searched a bit earlier and surprisingly didn't find an existing issue for the broader support.

**Comment by wlach** [comment]
For (1) and (2) I'd just use the exact same logic currently in use for ipynb (jupyter) notebooks, which AFAICT is this:

https://github.com/astral-sh/ruff/blob/cb2e277482f30c7b45dd3c5a163fef952d66281a/crates/ruff_linter/src/rules/flake8_bugbear/helpers.rs#L12

(3) and (4) are how I'd envision this working (as an enthusiastic user but non-developer of Marimo). Maybe I'll mention this on the Marimo discord and see if any of the devs there have thoughts.

**Comment by dylwil3** [comment]
> For (1) and (2) I'd just use the exact same logic currently in use for ipynb (jupyter) notebooks, which AFAICT is this:

I guess my question was more about how to translate that logic to Marimo, since I'm not super familiar with it. We can't literally apply that function, since as far as I understand a Marimo notebook is an ordinary Python file - we have no a priori notion of "cell" or cell offsets, etc. So I guess my main question is how to understand the notion of a "cell" in this setup. I can see that one part involves the decorator `@app.cell`, but I'm mainly confused about how to interpret returned values (does returning a value suppress displaying the last expression?), or whether there are any other differences that may be relevant here.

**Comment by wlach** [comment]
> I guess my question was more about how to translate that logic to Marimo, since I'm not super familiar with it. We can't literally apply that function, since as far as I understand a Marimo notebook is an ordinary Python file - we have no a priori notion of "cell" or cell offsets, etc. So I guess my main question is how to understand the notion of a "cell" in this setup. I can see that one part involves the decorator `@app.cell`, but I'm mainly confused about how to interpret returned values (does returning a value suppress displaying the last expression?), or whether there are any other differences that may be relevant here.

The return values (and function parameters) are hidden from the notebook view, so you're just left with the part of the function before the return statement in the notebook view. E.g. the notebook above translates to:

<img width="1436" height="897" alt="Image" src="https://github.com/user-attachments/assets/5d9193ea-5584-48fe-8a08-b88e8c37d995" />

https://molab.marimo.io/notebooks/nb_9zjTe3V6nrjb4F7TshETGk

So to answer your question, no, returning a value doesn't suppress the last expression. The rule is that marimo will always show the expression _before_ the return statement (mimicking the behaviour of jupyter, because the return values are abstracted away from the user when running inside the notebook view). See: https://docs.marimo.io/guides/outputs/

In reply to @ntBre, I personally don't really have any other feature requests for *ruff* wrt marimo other than this one. Mostly it just works, and catches real problems before they land.

**Comment by ozanozbeker** [comment]
All marimo notebooks start with something like:

```python
import marimo

__generated_with = "0.19.10"
app = marimo.App(width="full")
```

which could be a way to know it's a marimo notebook. Then the cells themselves are decorated with:

```python
@app.cell
```

**Comment by howsiyu** [comment]
I also like to disable the warning for last top-level expression in a jupyter code cell as in https://code.visualstudio.com/docs/python/jupyter-support-py
